File: src/sparkserve/model_lock.py
from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass
from pathlib import Path, PurePosixPath

_HEX40 = re.compile(r"[0-9a-f]{40}\Z")
_HEX64 = re.compile(r"[0-9a-f]{64}\Z")
_FORMATS = {"modelopt_nvfp4_safetensors", "gguf"}
_INVENTORIES = {"critical_metadata", "complete_checkpoint"}


@dataclass(frozen=True)
class LockedFile:
    path: str
    size: int
    sha256: str
    role: str


@dataclass(frozen=True)
class LockedModel:
    id: str
    repo: str
    revision: str
    architecture: str
    format: str
    quantization: str
    inventory: str
    checkpoint_bytes: int
    files: tuple[LockedFile, ...]


@dataclass(frozen=True)
class ModelLock:
    schema_version: int
    mirror: str
    models: tuple[LockedModel, ...]

    def model(self, model_id: str) -> LockedModel:
        matches = [model for model in self.models if model.id == model_id]
        if not matches:
            raise ValueError(f"model lock has no model {model_id!r}")
        return matches[0]


def _required(mapping: dict[str, object], key: str, expected: type) -> object:
    value = mapping.get(key)
    if not isinstance(value, expected):
        raise ValueError(f"{key} must be {expected.__name__}")
    return value


def _safe_relative_path(value: str) -> bool:
    path = PurePosixPath(value)
    return bool(value) and not path.is_absolute() and ".." not in path.parts


def parse_model_lock(payload: object) -> ModelLock:
    if not isinstance(payload, dict):
        raise ValueError("model lock root must be an object")
    schema_version = _required(payload, "schema_version", int)
    if schema_version != 1:
        raise ValueError(f"unsupported model lock schema {schema_version}")
    mirror = _required(payload, "mirror", str)
    if not mirror.startswith("https://"):
        raise ValueError("model mirror must use HTTPS")
    raw_models = _required(payload, "models", list)
    if not raw_models:
        raise ValueError("model lock must contain at least one model")

    models: list[LockedModel] = []
    seen_ids: set[str] = set()
    for raw_model in raw_models:
        if not isinstance(raw_model, dict):
            raise ValueError("model entry must be an object")
        model_id = _required(raw_model, "id", str)
        if not model_id or model_id in seen_ids:
            raise ValueError(f"model id must be non-empty and unique: {model_id!r}")
        seen_ids.add(model_id)
        revision = _required(raw_model, "revision", str)
        if not _HEX40.fullmatch(revision):
            raise ValueError(f"model {model_id} revision must be a 40-digit commit")
        model_format = _required(raw_model, "format", str)
        if model_format not in _FORMATS:
            raise ValueError(f"model {model_id} has unsupported format {model_format!r}")
        inventory = _required(raw_model, "inventory", str)
        if inventory not in _INVENTORIES:
            raise ValueError(f"model {model_id} has invalid inventory {inventory!r}")
        checkpoint_bytes = _required(raw_model, "checkpoint_bytes", int)
        if checkpoint_bytes <= 0:
            raise ValueError(f"model {model_id} checkpoint size must be positive")

        raw_files = _required(raw_model, "files", list)
        if not raw_files:
            raise ValueError(f"model {model_id} has no locked files")
        files: list[LockedFile] = []
        seen_paths: set[str] = set()
        for raw_file in raw_files:
            if not isinstance(raw_file, dict):
                raise ValueError(f"model {model_id} file entry must be an object")
            path = _required(raw_file, "path", str)
            if not _safe_relative_path(path) or path in seen_paths:
                raise ValueError(f"model {model_id} has unsafe or duplicate path {path!r}")
            seen_paths.add(path)
            size = _required(raw_file, "size", int)
            digest = _required(raw_file, "sha256", str)
            if size <= 0 or not _HEX64.fullmatch(digest):
                raise ValueError(f"model {model_id} has invalid integrity for {path}")
            files.append(
                LockedFile(
                    path=path,
                    size=size,
                    sha256=digest,
                    role=_required(raw_file, "role", str),
                )
            )
        if inventory == "complete_checkpoint" and sum(file.size for file in files) != checkpoint_bytes:
            raise ValueError(f"model {model_id} complete inventory does not match checkpoint size")
        models.append(
            LockedModel(
                id=model_id,
                repo=_required(raw_model, "repo", str),
                revision=revision,
                architecture=_required(raw_model, "architecture", str),
                format=model_format,
                quantization=_required(raw_model, "quantization", str),
                inventory=inventory,
                checkpoint_bytes=checkpoint_bytes,
                files=tuple(files),
            )
        )
    return ModelLock(schema_version=schema_version, mirror=mirror, models=tuple(models))


def load_model_lock(path: str | Path) -> ModelLock:
    with Path(path).open("rb") as handle:
        return parse_model_lock(json.load(handle))


def verify_model_files(model: LockedModel, model_root: str | Path) -> None:
    root = Path(model_root).resolve()
    for locked in model.files:
        path = (root / locked.path).resolve()
        if not path.is_relative_to(root):
            raise ValueError(f"locked path escapes model root: {locked.path}")
        try:
            stat = path.stat()
        except FileNotFoundError as error:
            raise ValueError(f"locked file is missing: {locked.path}") from error
        if stat.st_size != locked.size:
            raise ValueError(
                f"locked file size mismatch for {locked.path}: {stat.st_size} != {locked.size}"
            )
        digest = hashlib.sha256()
        with path.open("rb") as handle:
            for chunk in iter(lambda: handle.read(8 * 1024 * 1024), b""):
                digest.update(chunk)
        actual = digest.hexdigest()
        if actual != locked.sha256:
            raise ValueError(f"locked file SHA-256 mismatch for {locked.path}")

File: crates/sparkserve-runtime/src/model_lock.rs
use serde_json::{Map, Value};
use std::collections::BTreeSet;
use std::fmt::{Display, Formatter};
use std::fs::File;
use std::path::{Component, Path};

const REVISION_HEX_DIGITS: usize = 40;
const SHA256_HEX_DIGITS: usize = 64;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct LockedFile {
    pub path: String,
    pub size: u64,
    pub sha256: String,
    pub role: String,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct LockedModel {
    pub id: String,
    pub repo: String,
    pub revision: String,
    pub architecture: String,
    pub format: String,
    pub quantization: String,
    pub inventory: String,
    pub checkpoint_bytes: u64,
    pub files: Vec<LockedFile>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ModelLock {
    pub schema_version: u64,
    pub mirror: String,
    pub models: Vec<LockedModel>,
}

impl ModelLock {
    pub fn model(&self, id: &str) -> Result<&LockedModel, ModelLockError> {
        self.models
            .iter()
            .find(|model| model.id == id)
            .ok_or_else(|| ModelLockError::Invalid(format!("model lock has no model {id}")))
    }
}

#[derive(Debug)]
pub enum ModelLockError {
    Io(std::io::Error),
    Json(serde_json::Error),
    Invalid(String),
}

impl Display for ModelLockError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Io(error) => write!(formatter, "{error}"),
            Self::Json(error) => write!(formatter, "{error}"),
            Self::Invalid(message) => formatter.write_str(message),
        }
    }
}

impl std::error::Error for ModelLockError {}

impl From<std::io::Error> for ModelLockError {
    fn from(error: std::io::Error) -> Self {
        Self::Io(error)
    }
}

impl From<serde_json::Error> for ModelLockError {
    fn from(error: serde_json::Error) -> Self {
        Self::Json(error)
    }
}

pub fn load_model_lock(path: &Path) -> Result<ModelLock, ModelLockError> {
    parse_model_lock(serde_json::from_reader(File::open(path)?)?)
}

pub fn parse_model_lock(value: Value) -> Result<ModelLock, ModelLockError> {
    let root = object(&value, "model lock")?;
    let schema_version = unsigned(root, "schema_version")?;
    if schema_version != 1 {
        return invalid(format!("unsupported model lock schema {schema_version}"));
    }
    let mirror = string(root, "mirror")?;
    if !mirror.starts_with("https://") {
        return invalid("model mirror must use HTTPS");
    }
    let raw_models = array(root, "models")?;
    if raw_models.is_empty() {
        return invalid("model lock contains no models");
    }

    let mut ids = BTreeSet::new();
    let mut models = Vec::with_capacity(raw_models.len());
    for raw_model in raw_models {
        let model = parse_model(object(raw_model, "model entry")?)?;
        if !ids.insert(model.id.clone()) {
            return invalid(format!("duplicate model id {}", model.id));
        }
        models.push(model);
    }
    Ok(ModelLock {
        schema_version,
        mirror: mirror.to_owned(),
        models,
    })
}

fn parse_model(raw: &Map<String, Value>) -> Result<LockedModel, ModelLockError> {
    let id = string(raw, "id")?;
    if id.is_empty() {
        return invalid("model id must not be empty");
    }
    let revision = string(raw, "revision")?;
    if !lower_hex(revision, REVISION_HEX_DIGITS) {
        return invalid(format!("model {id} revision must be a 40-digit commit"));
    }
    let format = string(raw, "format")?;
    if !matches!(format, "modelopt_nvfp4_safetensors" | "gguf") {
        return invalid(format!("model {id} has unsupported format {format}"));
    }
    let inventory = string(raw, "inventory")?;
    if !matches!(inventory, "critical_metadata" | "complete_checkpoint") {
        return invalid(format!("model {id} has invalid inventory {inventory}"));
    }
    let checkpoint_bytes = unsigned(raw, "checkpoint_bytes")?;
    if checkpoint_bytes == 0 {
        return invalid(format!("model {id} checkpoint size must be positive"));
    }
    let raw_files = array(raw, "files")?;
    if raw_files.is_empty() {
        return invalid(format!("model {id} has no locked files"));
    }
    let mut paths = BTreeSet::new();
    let mut files = Vec::with_capacity(raw_files.len());
    for raw_file in raw_files {
        let file = parse_file(id, object(raw_file, "locked file")?)?;
        if !paths.insert(file.path.clone()) {
            return invalid(format!("model {id} has duplicate file {}", file.path));
        }
        files.push(file);
    }
    if inventory == "complete_checkpoint" {
        let inventory_bytes = files.iter().try_fold(0_u64, |total, file| {
            total.checked_add(file.size).ok_or_else(|| {
                ModelLockError::Invalid(format!("model {id} inventory size overflow"))
            })
        })?;
        if inventory_bytes != checkpoint_bytes {
            return invalid(format!(
                "model {id} inventory is {inventory_bytes} bytes, expected {checkpoint_bytes}"
            ));
        }
    }
    Ok(LockedModel {
        id: id.to_owned(),
        repo: string(raw, "repo")?.to_owned(),
        revision: revision.to_owned(),
        architecture: string(raw, "architecture")?.to_owned(),
        format: format.to_owned(),
        quantization: string(raw, "quantization")?.to_owned(),
        inventory: inventory.to_owned(),
        checkpoint_bytes,
        files,
    })
}

fn parse_file(model_id: &str, raw: &Map<String, Value>) -> Result<LockedFile, ModelLockError> {
    let path = string(raw, "path")?;
    if !safe_relative_path(path) {
        return invalid(format!("model {model_id} has unsafe file path {path}"));
    }
    let size = unsigned(raw, "size")?;
    let sha256 = string(raw, "sha256")?;
    if size == 0 || !lower_hex(sha256, SHA256_HEX_DIGITS) {
        return invalid(format!("model {model_id} has invalid integrity for {path}"));
    }
    Ok(LockedFile {
        path: path.to_owned(),
        size,
        sha256: sha256.to_owned(),
        role: string(raw, "role")?.to_owned(),
    })
}

fn object<'a>(value: &'a Value, label: &str) -> Result<&'a Map<String, Value>, ModelLockError> {
    value
        .as_object()
        .ok_or_else(|| ModelLockError::Invalid(format!("{label} must be an object")))
}

fn string<'a>(value: &'a Map<String, Value>, key: &str) -> Result<&'a str, ModelLockError> {
    value
        .get(key)
        .and_then(Value::as_str)
        .ok_or_else(|| ModelLockError::Invalid(format!("{key} must be a string")))
}

fn unsigned(value: &Map<String, Value>, key: &str) -> Result<u64, ModelLockError> {
    value
        .get(key)
        .and_then(Value::as_u64)
        .ok_or_else(|| ModelLockError::Invalid(format!("{key} must be an unsigned integer")))
}

fn array<'a>(value: &'a Map<String, Value>, key: &str) -> Result<&'a [Value], ModelLockError> {
    value
        .get(key)
        .and_then(Value::as_array)
        .map(Vec::as_slice)
        .ok_or_else(|| ModelLockError::Invalid(format!("{key} must be an array")))
}

fn lower_hex(value: &str, digits: usize) -> bool {
    value.len() == digits
        && value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
}

fn safe_relative_path(value: &str) -> bool {
    !value.is_empty()
        && Path::new(value)
            .components()
            .all(|component| matches!(component, Component::Normal(_)))
}

fn invalid<T>(message: impl Into<String>) -> Result<T, ModelLockError> {
    Err(ModelLockError::Invalid(message.into()))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn fixture() -> Value {
        json!({
            "schema_version": 1,
            "mirror": "https://hf-mirror.com",
            "models": [{
                "id": "glm",
                "repo": "owner/model",
                "revision": "0123456789abcdef0123456789abcdef01234567",
                "architecture": "GLM5NextForConditionalGeneration",
                "format": "gguf",
                "quantization": "UD-IQ3_XXS",
                "inventory": "complete_checkpoint",
                "checkpoint_bytes": 7,
                "files": [{
                    "path": "part-1.gguf",
                    "size": 7,
                    "sha256": "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
                    "role": "gguf_shard"
                }]
            }]
        })
    }

    #[test]
    fn parses_complete_inventory() {
        let lock = parse_model_lock(fixture()).expect("valid model lock");
        let model = lock.model("glm").expect("locked model");
        assert_eq!(model.checkpoint_bytes, 7);
        assert_eq!(model.files.len(), 1);
    }

    #[test]
    fn rejects_unsafe_path() {
        let mut value = fixture();
        value["models"][0]["files"][0]["path"] = Value::String("../part.gguf".into());
        let error = parse_model_lock(value).expect_err("unsafe path must fail");
        assert!(error.to_string().contains("unsafe file path"));
    }

    #[test]
    fn rejects_incomplete_inventory() {
        let mut value = fixture();
        value["models"][0]["checkpoint_bytes"] = Value::from(8);
        let error = parse_model_lock(value).expect_err("size mismatch must fail");
        assert!(error.to_string().contains("inventory is 7 bytes"));
    }
}
